R/utilities-simulation-results.R

```diff
@@ -0,0 +1,132 @@
+#' and \code{M} the number of individuals. If no individual contains the results for a given path, the value of the element is \code{NULL}
+#'
+#' @seealso \code{\link{getResultsValues}}
+#'
+#' @export
+#'
+#' @examples
+#' simPath <- system.file("extdata", "simple.pkml", package = "ospsuite")
+#' sim <- loadSimulation(simPath)
+#' simResults <- runSimulation(sim)
+#'
+#' paths <- c(
+#'   "Organism|PeripheralVenousBlood|Caffeine|Plasma (Peripheral Venous Blood)",
+#'   "Organism|ArterialBlood|Plasma|Caffeine|Concentration in container"
+#' )
+#'
+#' resultsValues <- getPopulationResultsValues(simResults, paths, individualIds = 0)
+#' # plot(resultsValues$`Organism|PeripheralVenousBlood|Caffeine|Plasma (Peripheral Venous Blood)`)
+getPopulationResultsValues <- function(simulationResults, quantitiesOrPaths, individualIds = NULL) {
+  validateIsOfType(simulationResults, "SimulationResults")
+  quantitiesOrPaths <- c(quantitiesOrPaths)
+  validateIsOfType(quantitiesOrPaths, c("Quantity", "character"))
+  validateIsNumeric(individualIds, nullAllowed = TRUE)
+
+  # If quantities are passed, get their paths.
+  paths <- quantitiesOrPaths
+  if (isOfType(quantitiesOrPaths, "Quantity")) {
+    paths <- lapply(quantitiesOrPaths, function(x) x$path)
+  }
+  paths <- unique(paths)
+
+  # If no specific individual ids are passed, iterate through all individuals
+  if (is.null(individualIds)) {
+    nrOfIndividuals <- simulationResults$count
+    # TODO: Can we assume that all individuals ids are ascending number from 0 to the number of individuals-1??
+    individualIds <- 0:(nrOfIndividuals - 1)
+  }
+  else {
+    individualIds <- unique(individualIds)
+    nrOfIndividuals <- length(individualIds)
+  }
+
+  # Pre-alloate a list with the number of entries equal to the number of paths
+  output <- vector("list", length(paths))
+  names(output) <- paths
+  # All time values are equal
+  timeVals <- simulationResults$timeValues
+
+  for (path in paths) {
+    # Pre-alloate a list with the number of entries equal to the number of individuals
+    outputValues <- vector("numeric")
+
+    for (individualId in individualIds) {
+      if (simulationResults$hasResultsForIndividual(individualId)) {
+        individualResult <- simulationResults$resultsForIndividual(as.integer(individualId))
+        # Check if the output exist. If not, the entry in the list of outputValues will have the value NULL
+        if (rClr::clrCall(individualResult, "HasValuesFor", path)) {
+          quantityValues <- rClr::clrCall(individualResult, "ValuesFor", path)
+          vals <- rClr::clrGet(quantityValues, "Values")
+          outputValues <- cbind(outputValues, vals, deparse.level = 0)
+        }
+      }
+    }
+    # Only fill the entry if at least one individual has the output
+    if (!(length(outputValues) == 0)) {
+      output[[path]][["x"]] <- timeVals
```

Comment:
Calling the entries x and y allows to simply use them in `plot(output$selectedPath)`.